### ISSUE TITLE
Check if throws

### ISSUE THREAD

**Issue by cmeeren** [hedgehogqa/fsharp-hedgehog#128]
Is it possible to check if a function throws an exception without having `try ... with` in the property and explicitly returning `true` or `false`?

**Comment by moodmosaic** [comment]
Would it be possible to provide an example?

**Comment by cmeeren** [comment]
```f#
[<Fact>]
let ``throws if hour is < 0 or > 23`` () =
  Property.check <| property {
    let! dt = Gen.dateTime
    let! hr = Range.linearBounded() |> Gen.int |> Gen.filter (fun hr -> hr < 0 || hr > 23)
    try
      nextOClock hr dt |> ignore
      return false
    with
      :? ArgumentException -> return true
  }
```

**Comment by cmeeren** [comment]
Ideally there'd be some kind of `Property.checkThrows<'exn>` that I could use. For example:

```f#
[<Fact>]
let ``throws if hour is < 0 or > 23`` () =
  Property.checkThrows<ArgumentException> <| property {
    let! dt = Gen.dateTime
    let! hr = Range.linearBounded() |> Gen.int |> Gen.filter (fun hr -> hr < 0 || hr > 23)
   nextOClock hr dt
```

**Comment by moodmosaic** [comment]
Oh, now I see what you mean; you *want* the exception to be thrown, and in this case, the property to pass.

**Comment by cmeeren** [comment]
👍

**Comment by moodmosaic** [comment]
@cmeeren, we recently added an internal function which does what you describe: https://github.com/hedgehogqa/fsharp-hedgehog/blob/70d445497cbba1af550ee45b07e6b85d5145f38d/src/Hedgehog/Property.fs#L382-L389

AFAICT, this function does exactly what you need. If so, we can make it public in the next release.

**Comment by cmeeren** [comment]
No, that does the opposite. It treats throws as failures. I want to test that a function does indeed throw an exception (the test should fail if nothing is thrown). Furthermore, it should check the type of exception (the test should fail if the wrong exception is thrown). And ideally, the user should be able to choose whether it needs to match the exception exactly or if a subclass of the specified exception is fine.

IIRC FsCheck does this nicely.

**Comment by moodmosaic** [comment]
OK, then it's the opposite of that function.

**Comment by cmeeren** [comment]
Yes. See my [example above](https://github.com/hedgehogqa/fsharp-hedgehog/issues/128#issuecomment-335400553) for a suggestion of a nice and short syntax:

```f#
Property.checkThrows<ArgumentException>
```

I suggest that the default behavior is that this passes if `ArgumentException` or any subclass of `ArgumentException` is thrown, and fails otherwise. If the user wants the test to pass for only `ArgumentException` (not a subclass), then I suggest another function:

```f#
Property.checkThrowsExact<ArgumentException>
```
